### PR TITLE
Develop game struct, it's api and tests.

### PR DESCRIPTION
**Pull request by vicentevieytes**
Introduces the Game struct:
```rust
struct Game {
    grid: Vec<Vec<Cell>>,
    alive_cells: Vec<Position>,
}
```
With the following API to create new instances and calculate the iterations of the game:
```rust
impl Game {
    pub fn from_cells(dimensions: Position, cells: &Vec<Position>) -> Result<Self, GameError>
    pub fn next(&mut self)
}
```

### CODE REVIEW

File: src/game.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    pub fn next(&mut self) {
+        let mut to_live = vec![];
+        let mut to_die = vec![];
+        //Iterate over all cells in the grid counting live_neighbors, cells which need to die or
+        //be given life are pushed to the respective vectors.
+        for (i, row) in self.grid.iter().enumerate() {
+            for (j, cell) in row.iter().enumerate() {
+                let live_neighbors = self.live_neighbors((i, j));
+
+                if cell.is_alive() {
```
Comment (MarcosNicolau) [inline src/game.rs]: You could move the rules calculation to its own function, it makes the loop kinda hard to follow

File: src/game.rs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+                    }
+                } else {
+                    //cell is dead
+                    if live_neighbors == 3 {
+                        to_live.push((i, j));
+                    }
+                }
+            }
+        }
+        self.kill_list(&to_die);
```
Comment (MarcosNicolau) [inline src/game.rs:64]: Can't you just mutate the cells in the previous loop, instead of creating two new lists and iterating over again?

File: src/game.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        }
+    }
+
+    fn give_life_list(&mut self, list: &Vec<Position>) {
+        for &(i, j) in list.iter() {
+            self.grid[i][j].give_life();
+        }
+    }
+
+    fn live_neighbors(&self, position: Position) -> i32 {
```
Comment (MarcosNicolau) [inline src/game.rs]: Look good

File: src/game.rs (hunk 1 above)
Comment (vicentevieytes) [reply src/game.rs]: Done, i abstracted this onto should_die and should_live functions which take a cell and it's number of neighbors

File: src/game.rs (hunk 2 above)
Comment (vicentevieytes) [reply src/game.rs:64]: Not really, at least not on that first loop.

I need to calculate all the ones that need to die or llive first, otherwise i could kill the alive neighbors that were needed to make the correct calculation.

Imagine this is the board where O is alive and X is dead:
O  O  O
O  X  X
the bottom middle element should come to life, but if i iterate from the top first and mutate while iterating i would end up killing all the neighbors before i can calculate the alive neighbors, leading it to not come to life.

**Review by MarcosNicolau — APPROVED** [review]
Great 😃
